### ISSUE TITLE
[flang] surprising warning with -Wunused-variable

### ISSUE THREAD

**Issue by ivan-pi** [llvm/llvm-project#187815]
For the snippet,

```fortran
subroutine test(n,a,b,c,d)
implicit none
integer, intent(in) :: n
real, intent(in) :: a(n), b(n)
real, intent(out) :: c(n), d(n)
integer :: i
do concurrent (i = 1:n)
    c(i) = a(i) + b(i)
    d(i) = a(i) - b(i)
end do
end subroutine
```

when compiling with `-pedantic` I get the warning,

```
/app/example.f90:6:12: warning: Value of local variable 'i' is never used [-Wunused-variable]
  integer :: i
```

If `integer :: i` is commented out, it gives however,

```
/app/example.f90:7:16: error: No explicit type declared for 'i'
  do concurrent (i = 1:n)
                 ^
```

No false warning is emitted, when the form,

```fortran
do concurrent (integer :: i = 1:n)
```
or a regular `DO` loop are used.
